### ISSUE TITLE
service registry as a built-in topic

### ISSUE THREAD

**Issue by budrus** [eclipse-iceoryx/iceoryx#415]
## Brief feature description

The discovery requests from user side are currently transferred via the message queue connection. This has some drawbacks

- The message queue is not only used for registration and creation of resources during startup but also during runtime
- If an application polls the discovery information with a high frequency the message queue becomes a bottleneck 
- We have a message size limitation on the message queue or unix domain socket, i.e. we need some transport protocol to split the response to several messages if there are a lot services. Currently it is a limitation as we do not have such a protocol 
- The `findService()` is very close to AUTOSAR ara::com and maybe cannot be used in a reasonable way in other frameworks 
- If someone wants to use ara::com service discovery callbacks, having only a polling based interface is a big drawback 

## Detailed information

The proposal is to have a service registry that provides it's content as a built-in topic, similar to the introspection topics. Then a subscriber could be created to get the information either in a polling way or event-driven by using the subscriber in combination with a waitset. The service registry publisher would use a history with size 1. Then every subscriber gets the current state on subscribe if desired. By providing the full CaPro ID information and not only service and instance IDs, this built-in topic could also be used for non AUTOSAR use cases. With this approach having a large service registry is fine as we don't have to send the content over the message queue and it's distributed with iceoryx' finest true zero-copy. I guess we can then also get rid of `getServiceRegistryChangeCounter()` which smells like workaround

## Todo
- [x] Create design decision document
- [x] Introduce poor mans `cxx::map` (Will be done in #829)
- [x] Refactor `ServiceDescription`
  - [x] Remove the two argument c'tor in `ServiceDescription`
  - [x] Remove three integers in `ServiceDescription`
  - [x] ~~Forbid creating Pub/Sub with `ServiceDescription(AnyServiceString, AnyInstanceString, "foo")` and forbid adding such a invalid `ServiceDescription` to the registry~~
  - [x] ~~Add [ServiceDescriptionBuilder](https://github.com/eclipse-iceoryx/iceoryx/pull/1047#discussion_r792652416)~~ Moved to #1141
- [x] Refactor `ServiceRegistry` to use the whole `ServiceDescription`
- [x] Error handling in `ServiceDescription(const cxx::Serialization& serial)` like `client_options.cpp`
- [x] Create example for new user-facing `DiscoveryInfo` class and move `getServiceRegistryChangeCounter`, `findService`, ~~`offerService` and `stopOfferService`~~ to this class 
  - [x] Extend `icediscovery`
  - [x] Write `README.md` for `icediscovery`
- ~~[ ] Introduce new [`StatusPorts`](https://github.com/eclipse-iceoryx/iceoryx/discussions/818)~~ Moved to #982
- [ ] ~~Built-in topic for service registry based on`BroadcastPorts`~~ Moved to #1135
   - [ ] ~~User facing API~~
- ~~[ ] Replace `InterfacePorts` with `StatusPorts`~~ Moved to #982
- [ ] ~~Mapping to Autosar Adaptive and APIs to work on an Autosar service level~~ Moved to #1103
- ~~[ ] Usage of #830~~
  - ~~[ ] Adaption of `ServiceRegistry::add()` (remove duplicate search)~~ (not needed, we basically have a multiset with counters)
- [x] `ServiceRegistry::remove()` should use same algorithm as `find()` to locate the entry (and then remove it)

Discussions:
- [x] Consider `ServiceRegistry::find("Foo", Wildcard, [] (auto& serviceDescription) { })` [API](https://github.com/eclipse-iceoryx/iceoryx/pull/860#discussion_r693374036)
- [x] Can `ServiceDescription::isValid()` be avoided?
- [x] When `ManyToManyPolicy` is used, would a user be interested in the number of offers? This is stored in the `ServiceRegistry` as `ServiceDescriptionEntry::referenceCounter` but currently not passed up to the user API.
- [ ] ~~Does the `ServiceRegistry` belong to `roudi` or `runtime`? Somehow, it sits in-between, currently it is in `source/roudi/service_registry.cpp`~~ Moved to #1200

Stretch goals:
- [ ] ~~Naming of the CaPro ID members, consider [comment](https://github.com/eclipse-iceoryx/iceoryx/pull/481/files#r553892569) from @ithier~~  Moved to #1200
- [ ] ~~Storage of IDs in service registry and ports (short IDs that are references to the string-based IDs, dictionary)~~ Moved to #1200
- [ ] ~~Decision on whether to remove wildcard in Cyclone DDS gateway~~ Moved to #1200
- [x] Check if `FindServiceError::UNABLE_TO_WRITE_TO_ROUDI_CHANNEL` can be removed
- [x] Check if `CaproMessageSubType` can be removed

**Comment by elBoberido** [comment]
A big drawback to the built-in topic is the worst case memory consumption. We maybe have to take into account that each application holds a different generation of the service registry sample, depending on how it's implemented. Even if we don't expose the sample to the user and do everything in the runtime, there might still be situation where multiple generations are held by various runtimes. With the introspection we currently assume that there is only one client and we just print some errors if the mempool runs out of chunks. It might be problematic to do the same with the service registry topic.

The proposed fix for #332 will enable segmented transmissions by IPOSIX PC channels. This doesn't solve the problem of the shared communication channel for registration, heartbeat and requesting resources like ports. I would split that into multiple channels, at least one just for registration/heartbeat and resource requests. This would still not solve the problem of sending large data by this channels but at least doesn't block registration. The IPC channel could be used as a fallback in case all chunks reserved for the service registry are exhausted. In order to make this work, the service registry must not be a built-in topic but still be requested by the IPC channel. So we would have the following scenario:
1. app requests the service registry data
2. RouDi has this information stored in a chunk and there is a spare chunk to use for new service registry data
    - the chunk is sent as relative pointer to the app by the IPC channel
    - the spare chunk is used as new storage for the service registry data
3. RouDi has this information stored in a chunk, but there is no spare chunk to use for service registry changes
    - RouDi serializes the data from the chunk and sends it to the app by the IPC channel

This ensures that the application always gets the information although it might be expensive if the chunks are exhausted. Alternatively instead of serializing the chunk, a pending response or error message could be sent.

**Comment by budrus** [comment]
good point @elBoberido. Having the MQ or UDS channel as point of contact and in the startup phase of an application is fine. But during runtime any possible interference from such a resource and potentially blocking calls on it should be avoided. I'm not sure if we solve this with your proposal without making things complicated. 
The "how many chunks would we need" question is valid. But before having a second request response channel via MQ I would try to solve it with iceoryx pub/sub or req/res
But I see that we need some more discussions

**Comment by elBoberido** [comment]
sure, iceoryx req/res would work exactly like the POSIX IPC channel, just that there would be a pending response/error response instead of serialization

**Comment by budrus** [comment]
When we touch the service registry here, should we also take care about naming of our IDs in the CaPro description? The proposal of @MatthiasKillat was to have Topics, Instances of Topics and Groups of Topics. I.e. renaming ServiceID to GroupID, InstanceID stays, renaming EventID to TopicID. 
Then we could define what capabilities our service registry has (e.g. get all instances of a specific topic, or get all topics and instances of a specific group. And for sure the documentation and mapping to ara::com, DDS and ROS
Sure we should come up with several PRs but maybe it makes sense to have it in this issue or would you prefer somehow linked individual issues?

@mossmaurice @MatthiasKillat

**Comment by mossmaurice** [comment]
@budrus I think it's a good idea to rename the three CaPro identifiers. We need to make sure to have good documentation, so that developers coming from different backgrounds feel at home. Ok from my side to do this in this issue.

**Comment by ithier** [comment]
With the proposed change to the composition of a service description, we would be restricting services to a single topic. Is this something that we want ?
Furthermore, I don't think it makes sense to have "multiple instances of a topic", which is a characteristic introduced by the proposal. To me, it makes more sense to have a different topic per instance.

Considering the above, it seems that the concepts of services and events are being conflated, and this makes it confusing (at least to me).

What are we trying to achieve with the change?
* ability to group together services ?
* a more "generic" abstraction ?

If the former, then I think just adding another component `groupId` would be sufficient.
If the latter, then I question why. I think the idea of services is pretty widespread and understandable. If wanting to go ahead with the different abstraction anyway, then I think the instance ID should be removed and instead be a part of the topic.
If both, then the proposal doesn't work and should be revised.

@budrus @MatthiasKillat

**Comment by budrus** [comment]
@mossmaurice It will be possible to attach the StatusPort to a Waitset or Listener to get notified if there are changes in the service registry, right? Use case would be to indicate with the offer state of a publisher thatf a specific service or functionality is available and to allow depending applications to get notified if there is a change in availability 
 
cc: @peddinavk

**Comment by mossmaurice** [comment]
> @mossmaurice It will be possible to attach the StatusPort to a Waitset or Listener to get notified if there are changes in the service registry, right? Use case would be to indicate with the offer state of a publisher thatf a specific service or functionality is available and to allow depending applications to get notified if there is a change in availability

Yep, that will be possible. For the user-facing `DiscoveryInfo` class it's planned to happen under the hood. `DiscoveryInfo` will just forward the attaching to the `StatusPort`, which lives in the `DiscoveryInfo` as a member. See more details in this [design decision document](https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/design/service-discovery.md#event-based-notification-alternative-d):

```
iox::popo::DiscoveryInfo discoveryInfo;
myListener.attachEvent(discoveryInfo, DATA_RECEIVED, createNotificationCallback(userDefinedCallback));
```

**Comment by mossmaurice** [comment]
Three staged approach as discussed by @FerdinandSpitzschnueffler @budrus @MatthiasKillat @mossmaurice 

Most important point is 1) to have a stable user API. Point 2) and 3) can be implemented underneath without changes to the API e.g. `v2.x`

1. `ServiceDiscovery` User API + `ServiceDisocvery` Example
   1. Extend `ServiceRegistry` with `std::multimap` for events @FerdinandSpitzschnueffler 
   2. Make `ServiceDiscovery` attachable to Listener and transmit change counter via a subscriber to be able to be notified on change
2. Finish `Prefixtree` @MatthiasKillat @elfenpiff 
   1. If time, exchange the `std::multimap` with `Prefixtree`, consider performance if `ServiceRegistry` is too large
   2. Instead of sending the change counter, send the complete `ServiceRegistry` via pub/sub
3. If time, finish `StatusPort` @mossmaurice 
   1. Instead of using pub/sub to send the `ServiceRegistry` send it via a `StatusPort` to avoid "denial of service" attack on RouDi (user subscribes to `ServiceRegistry` service and hold all chunks till we are out of memory)

**Comment by elfenpiff** [comment]
@mossmaurice If you need support I would like to write your examples to test the end user API since in my opinion it makes more sense to have someone look at the API who is not directly involved in the service discovery feature to look more from the end user perspective at this API.

**Comment by elBoberido** [comment]
@mossmaurice @FerdinandSpitzschnueffler while working on the tests for the server port, I noticed there is a `CaproMessageSubType` field in the `CaproMessage`. I somehow have the feeling this is obsolete and should be remove as part of this feature. What do you think?

**Comment by mossmaurice** [comment]
@elfenpiff Sure! A basic `icediscovery` example is already in #997 but wasn't merged to `master`. This could be a good basis (`icediscovery/iox_offer_service.cpp` and `icediscovery/iox_find_service.cpp`). It should explain:

1. How to setup a polling-based approach
1. How to setup a notification and callback based solution, explaining how one can use it to implement `ara::com`'s `StartFindService`

@elBoberido TBH, I currently don't know. It could be that `CaproMessageSubType` is related to the `InterfacePort`'s. I added an item to the list.

**Comment by MatthiasKillat** [comment]
@mossmaurice We cannot send the whole registry via pub sub, even if it is trivially copyable etc. It requires too much space. Not even the data itself, also the meta-data for ports (internal vectors of subscribers, service description etc.). The number of ports is limited. Hence even just sending the counter (a very small data type) causes us to require much more space than actually needed for ports.

This is less of a problem with a StatusPort since it does not depend on the maximum number of subscribers and hence less meta-data, but it is still very costly.

The most space efficient way is to have a condition variable (or interprocess semaphore) to wake up a discovery user when something changes. The user can then query the state of the discovery. While the API to do so is somewhat clear, the internal mechanism is not final. There are different possibilities with different advantages and implementation complexity (StatusPort, Request Response with Sockets or another IPC mechanism etc.). 

This mechanism can be improved incrementally later on, but the wake-up mechanism (e.g. condition variable) to inform us about changes needs to be in place in any case.

The search algorithm itself can also be improved later on, be it with prefix trees, maps or even another way still to be defined.
The choice of algorithm depends on the kind of queries and the number of services/events we want to support (if it is low, brute force may be good enough).

**Comment by MatthiasKillat** [comment]
The wake-up mechanism can work like this.
The user has a specific query Q in mind, e.g. wants to wait for a specific service to be available.
1. get the current counter C (this can be abstracted)
2. perform the query Q, if the query result is positive (service available) go to 6.
3. if Q is negative, get the counter again, if it is different from C go back to 1.
4. if the counter is the same C, issue a wait until the counter changes
5. when we wake up the counter will have changed (and so will the registry), hence we go back to 1.
6. query was positive

We need to enable the user to do just that with a somewhat nice API. We can hide some of the gory details, depending on how much freedom the user is supposed to have.

**Comment by elfenpiff** [comment]
To have some numbers I created a benchmark so that we have a discussion basis. The structure is like this:
1. Create a `cxx::vector` with 3 `capro::IdString which should reflect the service registry
2. The 3 `capro::IdString` have all the maximum size (completely filled)
3. The search service description is identical to all stored service descriptions inside the vector, except the last character of the event - this differs.

This means for every search run I have to iterate over the complete vector and have to perform 3 * SERVICE_REGISTRY_SIZE string comparisons. And the string comparison has to compare 100 characters every time. The absolute worst case.

Then I assume that the service registry has a size of 4096 (at the moment it is 512) and get the following numbers:
```
service registry size: 4096 identifier size: 100
Number of runs: 10000 time: 835.073ms, min/max search depth = 4096 / 4096 (should be equal)
Runtime required per run: 0.0835073ms
```

Or when I assume only 512 registered services:
```
service registry size: 512 identifier size: 100
Number of runs: 10000 time: 95.347ms, min/max search depth = 512 / 512 (should be equal)
Runtime required per run: 0.0095347ms
```

Here is the source code of the benchmark
```cpp
#include "iceoryx_hoofs/cxx/vector.hpp"
#include "iceoryx_posh/capro/service_description.hpp"

struct service_t
{
    iox::capro::IdString_t service;
    iox::capro::IdString_t instance;
    iox::capro::IdString_t event;
};

constexpr uint64_t MAX_NUMBER_OF_SERVICES = 512;
iox::cxx::vector<service_t, MAX_NUMBER_OF_SERVICES> services;

int main()
{
    iox::capro::IdString_t id(iox::cxx::TruncateToCapacity, std::string(iox::capro::IdString_t::capacity(), 'a'));
    iox::capro::IdString_t lastId(iox::cxx::TruncateToCapacity,
                                  std::string(iox::capro::IdString_t::capacity() - 1, 'a') + 'b');

    for (uint64_t i = 0; i < MAX_NUMBER_OF_SERVICES; ++i)
    {
        services.emplace_back(service_t{id, id, id});
    }

    /// search
    auto search = [&] {
        uint64_t i = 0;
        for (auto& s : services)
        {
            if (s.service == id && s.instance == id && s.event == lastId)
            {
                return i;
            }
            ++i;
        }
        return i;
    };

    constexpr uint64_t NUMBER_OF_RUNS = 10000;
    auto start = std::chrono::steady_clock::now();
    uint64_t maxSearchDepth = 0U, minSearchDepth = std::numeric_limits<uint64_t>::max();
    for (uint64_t i = 0; i < NUMBER_OF_RUNS; ++i)
    {
        maxSearchDepth = std::max(maxSearchDepth, search());
        minSearchDepth = std::min(minSearchDepth, search());
    }
    auto end = std::chrono::steady_clock::now();
    float runtimeInMs =
        static_cast<float>(std::chrono::duration_cast<std::chrono::microseconds>(end - start).count()) / 1000.0f;

    std::cout << "service registry size: " << services.size() << " identifier size: " << id.size() << std::endl;
    std::cout << "Number of runs: " << NUMBER_OF_RUNS << " time: " << runtimeInMs
              << "ms, min/max search depth = " << minSearchDepth << " / " << maxSearchDepth << " (should be equal)"
              << std::endl;
    std::cout << "Runtime required per run: " << runtimeInMs / NUMBER_OF_RUNS << "ms" << std::endl;
}
```

To quickly perform it yourself you can replace the code of `icedelivery/iox_publisher.cpp` with the code above and execute `./build/iceoryx_examples/icedelivery/iox-cpp-publisher`

FYI: @mossmaurice @MatthiasKillat @FerdinandSpitzschnueffler @elBoberido @budrus 

Now the question is, how fast do we want to be in the worst case? When I look at the iceperf output I see that the a UDS message has a latency around `0.0037ms` this means the back and forth communication has around `0.007ms`.

When the search is now much faster then `0.007ms` the user will never observe this performance increase since the UDS will throttle the whole process down. The simplest implementation has in the absolute worst case scenario a lookup time which is around `0.008ms` (for the current service registry size of 512). If we improve the speed further to `0.00001ms` the user has nevertheless a delay of `0.007ms`.

**Comment by MatthiasKillat** [comment]
@elfenpiff It should also be considered that removing entries from vectors is costly.

**Comment by MatthiasKillat** [comment]
@elfenpiff I modified the benchmark slightly, but yes worst case look-up for say 2000 services is below 1 ms. This does not include constructing the result (in this case the search is unsuccessful anyway). But creating the resultset must be done for all methods, so this can be factored out.

```
#include "iceoryx_hoofs/cxx/vector.hpp"
#include "iceoryx_posh/capro/service_description.hpp"


constexpr uint64_t MAX_NUMBER_OF_SERVICES = 2000;
constexpr uint64_t NUMBER_OF_RUNS = 100000;
constexpr uint64_t ID_LENGTH = 100;


using string_t = iox::cxx::string<ID_LENGTH>;
struct service_t
{
    string_t service;
    string_t instance;
    string_t event;
};


iox::cxx::vector<service_t, MAX_NUMBER_OF_SERVICES> services;

int main()
{
    string_t id(iox::cxx::TruncateToCapacity, std::string(string_t::capacity(), 'a'));

    string_t nonexistingId(iox::cxx::TruncateToCapacity, std::string(string_t::capacity() - 1, 'a') + 'b');

    for (uint64_t i = 0; i < MAX_NUMBER_OF_SERVICES; ++i)
    {
        services.emplace_back(service_t{id, id, id});
    }

    /// search
    auto search = [&] {
        uint64_t i = 0;
        for (auto& s : services)
        {
            if (s.service == id && s.instance == id && s.event == nonexistingId)
            {
                return i;
            }
            ++i;
        }
        return i;
    };

    auto start = std::chrono::high_resolution_clock::now();
    for (uint64_t i = 0; i < NUMBER_OF_RUNS; ++i)
    {
        search();
    }
    auto end = std::chrono::high_resolution_clock::now();

    float runtimeInMs =
        static_cast<float>(std::chrono::duration_cast<std::chrono::microseconds>(end - start).count()) / 1000.0f;

    std::cout << "service registry size: " << services.size() << " identifier size: " << id.size() << std::endl;
    std::cout << "Number of runs: " << NUMBER_OF_RUNS << " time: " << runtimeInMs << "ms" << std::endl;
    std::cout << "Runtime required per run: " << runtimeInMs / NUMBER_OF_RUNS << "ms" << std::endl;
}
```

**Comment by elfenpiff** [comment]
@budrus @MatthiasKillat @FerdinandSpitzschnueffler @mossmaurice   In my opinion the best solution would be:
1. Use publish subscribe (as temporary solution since it is resource hungry) and couple `MAX_PROCESS_NUMBER` in posh types with `MAX_SUBSCRIBERS_PER_PUBLISHER` and reduce the `MAX_PROCESS_NUMBER` to 256.
2. Create a memory pool inside the management segment with around 10 chunks of size `cxx::vector<StringTriple>` and use it for the publisher in the registry.
3. The publisher sends the whole service registry as `cxx::vector<StringTriple>` to the subscribers on every update. History Size = 1, Queue Size = 1
4. The serviceDiscovery copies this vector and performs all search operations on it. Push notification can be realized via the subscriber.

When this is running we can write a benchmark (one process with a single process roudi) and optimize the whole thing.

**Comment by elBoberido** [comment]
I'm not sure whether I want to join this discussion but here my 2 cents for a another option:
- plain array with `cxx::string` or `ServiceDiscovery` + ABA counter + flag to indicate whether the index is currently updating  in shared memory
- `PrefixTree` for fast lookup if we need the performance -> give us the indices in the array in shm
-  send the matched services as index + ABA counter via UDS to the client; additionally the service registry change counter should be sent
- client copies the items from all indices and checks whether the ABA counter did not change and the `updating` flag is not set
  - if the ABA counter of one item changed -> repeat the request
  - it one item has the `updating` flag -> repeat the request
  - if the received service registry change counter differs from the one in SHM -> repeat the request

Since we have limited amount of stack on QNX (by default 512kB for the main thread and  only128kb for further threads), one does not even need to copy all items but can process them one by one

Just my 2 cents

cc @budrus @FerdinandSpitzschnueffler @MatthiasKillat @mossmaurice @elfenpiff

**Comment by budrus** [comment]
@elBoberido Sounds like the most advanced solution we can identify ;-)

Another question, if the `findService()` returns an `InstanceContainer`, and the user searches for (*,*,*) would we not always end up in having all the CapProIDs in one data structure on the stack?

**Comment by budrus** [comment]
OK as we have Friday and everyone is invited to come up with crazy ideas.

Current challenges

1. UDS has no fragmentation today
2. With pub/sub we have the challenge that you have to either hold the last message or copy the last message. We need this to be able to have another search request also when there was no update in between and the queue is empty. This could lead to either the worst case that every application has another chunk and we need a lot of them or to an object of 0.5 - 1MB on the stack
3. The status port has the problem that we also have to do a temporary copy to ensure that we do not operate on a corrupted data structure. Which would also live on the stack 
4. For @elBoberido's proposal I first have to go back to school 

So what about this 

- We fix #938 
- We a publisher for the service registry in RouDi with history capacity size 1 
- We use a subscriber in the `ServiceDiscovery` with queue size 1 and history request 1
- When we call a `findService()` we check if the subscriber has a new sample in the queue. When using a waitset or listener this will be the case as this new sample triggered
- When there is no new sample in the queue, we make a placement new of the subscriber. The old one will be deleted and the new one created. This is a UDS request comparable to having a `findService()` call over the UDS. When the c'tor returns we'll have a subscriber with the latest service registry in the queue
- We do not have to keep a copy of the service registry in the `ServiceDiscovery`
- We can search on the sample taken from the queue fill the `InstanceContainer` and return the loan

@mossmaurice @elfenpiff @elBoberido @MatthiasKillat @FerdinandSpitzschnueffler

**Comment by elBoberido** [comment]
@budrus if `findService` returns a `cxx::vector<ServiceDescription, MAX_SERVICES>` with `MAX_SERVICES = 1000`, then it does not matter whether we get one or all services. Since the vector always requires the full size on the stack. With QNX this would be a problem. If a functional approach is used, e.g.
```cpp
findService([&] (const auto& ServiceDescription s) {
    // do stuff with the ServiceDescription
});
```
We shift prevent this problem and the user has to decide what to do, e.g.
```cpp
iox::cxx::vector<ServiceDescription, 10> v;
findService([&] (const auto& ServiceDescription s) {
    if(!v.push(s)) {
        // handle error
    }
});
```
or just
```cpp
findService([&] (const auto& ServiceDescription s) {
    if(s.getServiceIDString == "Foo") {
        // react on Foo
    } else if (s.getServiceIDString == "Bar") {
        // react on Bar
    } else {
        // it must be friday, why did I look for a service I'm not interested in
    }
});
```

**Comment by budrus** [comment]
@elBoberido 

> if findService returns a cxx::vector<ServiceDescription, MAX_SERVICES> with MAX_SERVICES = 1000, then it does not matter whether we get one or all services

True, my bad

**Comment by elfenpiff** [comment]
> ```c++
> findService(/*search parameter */, [&] (const auto& ServiceDescription s) {
>     if(s.getServiceIDString == "Foo") {
>         // react on Foo
>     } else if (s.getServiceIDString == "Bar") {
>         // react on Bar
>     } else {
>         // it must be friday, why did I look for a service I'm not interested in
>     }
> });
> ```

I am in favor of this approach. Yes it is a functional one which is maybe not mainstream C++ but we have very good arguments for it and the use case and performance tell us that this is the right approch.

**Comment by MatthiasKillat** [comment]
@elBoberido I currently change the internal implementation to vector. I will have a pull request today where I change the t the implementation to use a `vector` but do NOT change the interface. This eliminates the `map`.

Afterwards we can change the whole interface to a functional approach but I prefer not to do this in one go since it affects tests and other classes and needs careful consideration. There are other considerations on how to communicate the result efficiently, @elBoberido sketched potential approaches to send only indices etc. but this requires some more thought, it is not that simple.

I was also measuring performance of the tree approach (on a 5000 entry registry), but am not done yet. Preliminary results show it is much faster in the average cases but there are pathological cases for all searches where essentially the (intermediate) result is large. In this case it cannot perform (much) better than brute force search since in a way it also searches the whole tree (as the result is that large). It performs well if the result is small, which is the case on average. I will have numbers later.

**Comment by mossmaurice** [comment]
@MatthiasKillat
> This mechanism can be improved incrementally later on, but the wake-up mechanism (e.g. condition variable) to inform us about changes needs to be in place in any case, so it should be considered next.

I'll be working on it this week.

@elfenpiff Thanks a bunch for creating the benmchark! Let's first opt for the `cxx::vector`.

@elBoberido
> Since we have limited amount of stack on QNX (by default 512kB for the main thread and only128kb for further threads),
> With QNX this would be a problem

I saw stack size usages of several MB in ADAS systems. Hence, a developer building an ADAS or robotics system using QNX will need to change the default stack size anyway. However, I still like the functional approach a lot, so let's go for it :)

**Comment by MatthiasKillat** [comment]
Some preliminary numbers, I cannot quickly make the code available (different branches etc.)
- full 5000 entry registry
- partially random keys (needed, duplicates are good since they do not fill the registry) -> simulates some kind of average case
- entries have maximum size (100)and overlap in service and instance, which makes it bad for brute force but also for "dumb" tree searches without heuristics to reduce the result size quickly (these are in place)
- events are random (I imagine they will be fairly diverse in reality, so this is realistic)
- exact search for one triple which exists only once


This is not exactly the worst case but I consider it a lot worse then "average" (average is tricky business and cannot be defined without an input distribution, so this remains vague).

For 100 000 searches for each variant we have:

1. current map implementation: 31783.7 ms , average search time 0.317837 ms
2. vector implementation with linear search: 8174.22 ms, average search time  0.0817422 ms
3. triple prefixtree implementation (with intersection heuristics): 75.5096 ms, average search time 0.000755096 ms

All run the same benchmark setup, and it is worse than the expected case but not unrealistic.

The major downside of the tree is 1) code complexity (but this is true for all elaborate search structures, i.e. would be for our own map as well) and 2) memory consumption. The advantage is search time and this would only become more pronounced for larger numbers of entries (should not grow by much, compared to e.g. brute force)

A 5000 entry registry with trees is about 64 MByte, for 1000 entries we still need 13 MByte (it scales close to linearly with some constant which we can derive from a few data points). This is essentially the price we pay for speed and a static memory model (n the dynamic memory case it could still grow that large, but on average it would not).

**Comment by mossmaurice** [comment]
User API finished, implementation will be refactored with #1200
